### ISSUE TITLE
Using ,= on a SetHash causes an infinite loop

### ISSUE THREAD

**Issue by nikola-benes** [rakudo/rakudo#4678]
## The Problem

```raku
my %s is SetHash;
%s ,= 1;
```

## Expected Behavior

Not quite sure, but definitely not an infinite loop.

## Actual Behavior

Seems to be an infinite loop. The raku process takes 100% CPU and slowly eats all available memory.

## Environment
* Operating system: Linux
* Compiler version (`perl6 -v` or `raku -v`): v2021.08

**Comment by lizmat** [comment]
Actually, this is **not** the problem:
````raku
my %s is SetHash;
%s ,= 1;
````
but it most likely does not do what you think it does.  `%s ,= 1` is short for `%s = %s,1`.  Which means you re-initialize the `SetHash` with **ITSELF** and the value `1`.
````raku
my %s is SetHash;
%s ,= 1;
dd %s{1, %s};  # (Bool::True, Bool::True)
````
The problem starts when trying to create a representation of the resulting `SetHash`.

I'm not sure how to fix this, am in two minds about this.  The options I see are:
1. Put in extra checks in stringification / .rakufication of `SetHash` (and by consequence also of `BagHash` and `MixHash`)
2. Do **NOT** allow the `SetHash` become an element of itself, with an error describing that you could use `.set` on SetHashes to add elements to the `SetHash`.
3. Make an occurrence of itself automatically refer to the keys of the `SetHash`, thereby getting closer to DWIM, with the danger of a big WAT.

I'm leaning towards 2 at the moment, as I feel has the right balance between performance and making the developer use the correct way of doing this (aka, using the `.set` method).

**Comment by lizmat** [comment]
@jdv This is **NOT** a blocker.

**Comment by jnthn** [comment]
I lean towards option 1, on the basis that I think the performance of stringification or rakuification of `SetHash` and friends is an unlikely hot path (I suspect it's mostly use for debug output during development), and also because option 1 would fix the problem in general (including non-immediate nesting), whereas 2 likely would not.

**Comment by lizmat** [comment]
Yeah, but why would people do `%s ,= 1` ??   I can only think they assume that `%s` will flatten its keys.

Sanity check: @nikola-benes What did you think to achieve with `%s ,= 1` ??

**Comment by nikola-benes** [comment]
Well, with `Hash`es `,=` simply adds a new pair to the hash:

```raku
my %h;
%h ,= :a(1);
%h ,= :b(2);
say %h  # {a => 1, b => 2}
```

I was just curious whether this also works for `SetHash`es. (I am still trying to grasp the rules of automagic flattening…)

**Comment by lizmat** [comment]
Right.  `QuantHashes` are not `Iterable`, whereas `Hash`es are:
````raku
say Hash ~~ Iterable;  # True
say SetHash ~~ Iterable;  # False
````
So, a `SetHash` (well, basically **any** `QuantHash`) on the RHS, will **NOT** flatten.

Perhaps we need a 4th option:
- Fix stringification (which also includes WHICHification) **AND** issue a warning that it probably isn't what you wanted.

**Comment by MasterDuke17** [comment]
> Right. `QuantHashes` are not `Iterable`, whereas `Hash`es are:

Huh, I didn't know that before. Why aren't they `Iterable`?

**Comment by lizmat** [comment]
If I remember correctly, it's because TimToady thought that `QuantHash`es should always be considered as an item, regardless of context.

**Comment by nikola-benes** [comment]
The more I try to understand this language, the more confused I get… If `Set`s are not iterable, why does `.say for set(1, 2, 3);` print three separate lines? Or is the `Iterable` role not connected to the concept of iteration in a `for` cycle?

**Comment by lizmat** [comment]
Well, you could argue that Raku has its warts.  :-)

When you do `for` on something, you are indicating you **want** to iterate over it.
This means it will call the `iterator` method on that something.  And every class has an (inherited) `iterator` method:
````raku
dd 42.iterator;  # Rakudo::Iterator::ReifiedListIterator.new
````
This allows you to say:
````raku
.say for 42;  # 42
````
`QuantHash`es also have `iterator` methods, that will produce `Pair`s of key and value of its elements.

Because `List`s *are* `Iterable`, you can say something like:
````raku
my %h = :42a, :666b, :137c;
say %h<a b c>;  # (42 666 137)
````
because the `postcircumfix:<{ }>` candidate sees that the argument (a `List` with "a", "b", "c") is `Iterable`, and thus starts iterating over it to produce values.

Does that make sense to you, @nikola-benes ?

**Comment by lizmat** [comment]
Tests added with https://github.com/Raku/roast/commit/1b14e5eb27
